Source/Scene/ModelInstanceCollection.js

```diff
@@ -0,0 +1,675 @@
+            var instancedUniforms = getInstancedUniforms(collection, programName);
+            for (var uniform in instancedUniforms) {
+                if (instancedUniforms.hasOwnProperty(uniform)) {
+                    uniformMap[uniform] = undefined;
+                }
+            }
+
+            return uniformMap;
+        };
+    }
+
+    function getPickUniformMapCallback() {
+        return function(uniformMap) {
+            uniformMap.czm_pickColor = undefined;
+            return uniformMap;
+        };
+    }
+
+    var instanceMatrix = new Matrix4();
+
+    function updateVertexBuffer(collection, context) {
+        if (!collection._instancingSupported) {
+            return;
+        }
+
+        var typedArray = collection._typedArray;
+        var vertexBuffer = collection._vertexBuffer;
+        var createVertexBuffer = collection._createVertexBuffer;
+
+        var instancesLength = collection.instancesLength;
+        var dynamic = collection._dynamic;
+        var viewMatrix = context.uniformState.view;
+        var center = dynamic ? Cartesian3.ZERO : collection._boundingSphere.center;
+
+        if (createVertexBuffer) {
+            typedArray = new Float32Array(instancesLength * 16);
+        }
+
+        for (var i = 0; i < instancesLength; ++i) {
+            var instance = collection._instances[i];
+            var modelMatrix = instance.modelMatrix;
+            var pickColor = defined(instance.pickId) ? instance.pickId.color : Color.WHITE;
+
+            if (dynamic) {
+                Matrix4.multiplyTransformation(viewMatrix, modelMatrix, instanceMatrix);
+            } else {
+                instanceMatrix = modelMatrix;
+            }
+
+            // First three rows of the model matrix
+            typedArray[i * 16 + 0] = instanceMatrix[0];
+            typedArray[i * 16 + 1] = instanceMatrix[4];
+            typedArray[i * 16 + 2] = instanceMatrix[8];
+            typedArray[i * 16 + 3] = instanceMatrix[12] - center.x;
+            typedArray[i * 16 + 4] = instanceMatrix[1];
+            typedArray[i * 16 + 5] = instanceMatrix[5];
+            typedArray[i * 16 + 6] = instanceMatrix[9];
+            typedArray[i * 16 + 7] = instanceMatrix[13] - center.y;
+            typedArray[i * 16 + 8] = instanceMatrix[2];
+            typedArray[i * 16 + 9] = instanceMatrix[6];
+            typedArray[i * 16 + 10] = instanceMatrix[10];
+            typedArray[i * 16 + 11] = instanceMatrix[14] - center.z;
+
+            // Other instance data like pickColor, color, etc. Colors can be packed like in BillboardCollection as needed.
+            typedArray[i * 16 + 12] = pickColor.red;
+            typedArray[i * 16 + 13] = pickColor.green;
+            typedArray[i * 16 + 14] = pickColor.blue;
+            typedArray[i * 16 + 15] = pickColor.alpha;
+        }
+
+        if (createVertexBuffer) {
+            var usage = dynamic ? BufferUsage.STREAM_DRAW : BufferUsage.STATIC_DRAW;
```

Comment:
Style-wise, we don't need to declare since temporary since we only use it once.  It is fine and still clean to put the expression in the object literal below.